File: nexus-reader/src/utils/interfaces.ts
/**
 * Standardized Interfaces for Nexus Reader Components
 * Provides loose coupling and high cohesion through standardized contracts
 */

// ===== Core Interfaces =====

export interface BookSourceEngine {
  readonly name: string
  readonly version: string

  supportsUrl(url: string): boolean

  searchBooks(query: string, page?: number): Promise<BookItem[]>
  getBookDetails(url: string): Promise<BookItem>
  getTableOfContents(url: string): Promise<TocItem[]>
  getChapterContent(url: string): Promise<Chapter>

  testConnectivity(): Promise<boolean>
  getHealthStatus(): EngineHealthStatus
  getStatistics(): EngineStatistics
}

export interface Fetcher {
  get(url: string, headers?: Record<string, string>): Promise<FetchResponse>
  post(url: string, body: string, headers?: Record<string, string>): Promise<FetchResponse>
  getStatistics(): FetcherStatistics
}

export interface Cache<K = string, V = any> {
  get(key: K): Promise<V | null>
  set(key: K, value: V, ttlSeconds?: number): Promise<void>
  delete(key: K): Promise<boolean>
  clear(): Promise<void>
  getStatistics(): CacheStatistics
}

export interface Storage {
  store(key: string, data: Uint8Array): Promise<void>
  retrieve(key: string): Promise<Uint8Array | null>
  delete(key: string): Promise<boolean>
  listKeys(prefix?: string): Promise<string[]>
  getStatistics(): StorageStatistics
}

export interface ConfigProvider {
  get<T>(key: string, defaultValue?: T): T
  set<T>(key: string, value: T): void
  watch<T>(key: string, callback: (value: T) => void): () => void
  getAll(): Record<string, any>
  validate(): ValidationResult
}

export interface HealthMonitor {
  recordSuccess(operation: string, duration: number, metadata?: Record<string, any>): void
  recordFailure(operation: string, error: Error, metadata?: Record<string, any>): void
  getHealthStatus(): HealthStatus
  getStatistics(): HealthStatistics
}

export interface MetricsCollector {
  incrementCounter(name: string, value?: number, labels?: Record<string, string>): void
  setGauge(name: string, value: number, labels?: Record<string, string>): void
  observeHistogram(name: string, value: number, labels?: Record<string, string>): void
  getMetrics(): Record<string, MetricValue>
}

// ===== Data Structures =====

export interface ValidationResult {
  isValid: boolean
  errors: string[]
  warnings: string[]
}

export interface EngineHealthStatus {
  status: HealthState
  lastCheck: Date
  responseTime?: number
  errorCount: number
  successCount: number
}

export interface EngineStatistics {
  totalRequests: number
  successfulRequests: number
  failedRequests: number
  averageResponseTime: number
  uptime: number
  memoryUsage?: number
}

export interface FetcherStatistics {
  totalRequests: number
  successfulRequests: number
  failedRequests: number
  totalBytesDownloaded: number
  averageResponseTime: number
  activeConnections: number
}

export interface CacheStatistics {
  totalEntries: number
  hitCount: number
  missCount: number
  evictionCount: number
  hitRate: number
  totalSizeBytes: number
}

export interface StorageStatistics {
  totalFiles: number
  totalSizeBytes: number
  readOperations: number
  writeOperations: number
  deleteOperations: number
  averageReadTime: number
  averageWriteTime: number
}

export interface HealthStatus {
  overallState: HealthState
  components: Record<string, ComponentHealth>
  lastCheck: Date
}

export interface ComponentHealth {
  state: HealthState
  responseTime?: number
  errorMessage?: string
  lastSuccess?: Date
  lastFailure?: Date
}

export interface HealthStatistics {
  totalChecks: number
  successfulChecks: number
  failedChecks: number
  averageResponseTime: number
  uptimePercentage: number
}

export interface ConfigChangeEvent<T = any> {
  key: string
  oldValue?: T
  newValue: T
  source: string
  timestamp: Date
}

export type MetricValue =
  | { type: 'counter'; value: number }
  | { type: 'gauge'; value: number }
  | { type: 'histogram'; value: number[] }

// ===== Enumerations =====

export enum HealthState {
  HEALTHY = 'healthy',
  DEGRADED = 'degraded',
  UNHEALTHY = 'unhealthy',
  UNKNOWN = 'unknown'
}

// ===== Domain Types =====

export interface BookItem {
  id: string
  title: string
  author: string
  description?: string
  cover?: string
  url: string
  source: string
  tags?: string[]
  status?: string
  wordCount?: number
  updateTime?: Date
}

export interface TocItem {
  id: string
  title: string
  url: string
  index: number
  isVip?: boolean
  updateTime?: Date
}

export interface Chapter {
  id: string
  title: string
  content: string
  url: string
  index: number
  wordCount?: number
  updateTime?: Date
}

export interface FetchResponse {
  status: number
  headers: Record<string, string>
  body: string
  url: string
}

// ===== Factory Interfaces =====

export interface EngineFactory {
  createEngine(engineType: string, config?: any): BookSourceEngine
  getAvailableEngines(): string[]
}

export interface CacheFactory {
  createCache(cacheType: string, config?: any): Cache
  getAvailableCaches(): string[]
}

export interface ServiceFactory {
  createService<T>(serviceType: string, config?: any): T
  getAvailableServices(): string[]
}

// ===== Abstract Base Classes =====

export abstract class BaseBookSourceEngine implements BookSourceEngine {
  protected _stats: EngineStatistics
  protected _startTime: Date

  constructor(public readonly name: string, public readonly version: string = '1.0.0') {
    this._startTime = new Date()
    this._stats = {
      totalRequests: 0,
      successfulRequests: 0,
      failedRequests: 0,
      averageResponseTime: 0,
      uptime: 0,
      memoryUsage: undefined
    }
  }

  abstract supportsUrl(url: string): boolean
  abstract searchBooks(query: string, page?: number): Promise<BookItem[]>
  abstract getBookDetails(url: string): Promise<BookItem>
  abstract getTableOfContents(url: string): Promise<TocItem[]>
  abstract getChapterContent(url: string): Promise<Chapter>

  async testConnectivity(): Promise<boolean> {
    try {
      // Default implementation - can be overridden
      return true
    } catch {
      return false
    }
  }

  getHealthStatus(): EngineHealthStatus {
    const successRate = this._stats.totalRequests > 0
      ? this._stats.successfulRequests / this._stats.totalRequests
      : 0

    let status = HealthState.HEALTHY
    if (successRate < 0.5) status = HealthState.UNHEALTHY
    else if (successRate < 0.8) status = HealthState.DEGRADED

    return {
      status,
      lastCheck: new Date(),
      responseTime: this._stats.averageResponseTime,
      errorCount: this._stats.failedRequests,
      successCount: this._stats.successfulRequests
    }
  }

  getStatistics(): EngineStatistics {
    this._stats.uptime = Date.now() - this._startTime.getTime()
    return { ...this._stats }
  }

  protected recordRequest(success: boolean, duration: number): void {
    this._stats.totalRequests++
    if (success) {
      this._stats.successfulRequests++
    } else {
      this._stats.failedRequests++
    }

    // Update rolling average response time
    const alpha = 0.1
    this._stats.averageResponseTime =
      this._stats.averageResponseTime * (1 - alpha) + duration * alpha
  }

  async shutdown(): Promise<void> {
    // Optional cleanup
  }
}

export abstract class BaseCache<K = string, V = any> implements Cache<K, V> {
  protected _stats: CacheStatistics

  constructor() {
    this._stats = {
      totalEntries: 0,
      hitCount: 0,
      missCount: 0,
      evictionCount: 0,
      hitRate: 0,
      totalSizeBytes: 0
    }
  }

  abstract get(key: K): Promise<V | null>
  abstract set(key: K, value: V, ttlSeconds?: number): Promise<void>
  abstract delete(key: K): Promise<boolean>
  abstract clear(): Promise<void>

  getStatistics(): CacheStatistics {
    const total = this._stats.hitCount + this._stats.missCount
    this._stats.hitRate = total > 0 ? this._stats.hitCount / total : 0
    return { ...this._stats }
  }

  protected recordHit(): void {
    this._stats.hitCount++
  }

  protected recordMiss(): void {
    this._stats.missCount++
  }

  protected recordEviction(): void {
    this._stats.evictionCount++
  }
}

// ===== Utility Functions =====

export function createServiceFactory<T>(): ServiceFactory {
  const services = new Map<string, () => T>()

  return {
    createService(serviceType: string, config?: any): T {
      const factory = services.get(serviceType)
      if (!factory) {
        throw new Error(`Unknown service type: ${serviceType}`)
      }
      return factory()
    },

    getAvailableServices(): string[] {
      return Array.from(services.keys())
    },

    registerService: (serviceType: string, factory: () => T) => {
      services.set(serviceType, factory)
    }
  } as ServiceFactory & { registerService: (type: string, factory: () => T) => void }
}

File: nexus-lite/nexus-core/src/interfaces.rs
//! Standardized Interfaces for Nexus Components
//!
//! This module defines the core interfaces that all Nexus components must implement
//! to ensure loose coupling and high cohesion.

use crate::error::EngineError;
use crate::types::{BookItem, Chapter, FetchResponse, TocItem};
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Standard interface for book source engines
#[async_trait]
pub trait BookSourceEngine: Send + Sync {
    /// Get the engine name
    fn name(&self) -> &str;

    /// Get the engine version
    fn version(&self) -> &str;

    /// Check if the engine supports a given URL
    fn supports_url(&self, url: &str) -> bool;

    /// Search for books
    async fn search_books(
        &self,
        query: &str,
        page: Option<u32>,
    ) -> Result<Vec<BookItem>, EngineError>;

    /// Get book details
    async fn get_book_details(&self, url: &str) -> Result<BookItem, EngineError>;

    /// Get table of contents
    async fn get_table_of_contents(&self, url: &str) -> Result<Vec<TocItem>, EngineError>;

    /// Get chapter content
    async fn get_chapter_content(&self, url: &str) -> Result<Chapter, EngineError>;

    /// Test the engine connectivity
    async fn test_connectivity(&self) -> Result<(), EngineError>;

    /// Get engine health status
    fn health_status(&self) -> EngineHealthStatus;

    /// Get engine statistics
    fn statistics(&self) -> EngineStatistics;
}

/// Standard interface for fetchers (HTTP clients)
#[async_trait]
pub trait Fetcher: Send + Sync {
    /// Execute GET request
    async fn get(
        &self,
        url: &str,
        headers: Option<HashMap<String, String>>,
    ) -> Result<FetchResponse, EngineError>;

    /// Execute POST request
    async fn post(
        &self,
        url: &str,
        body: &str,
        headers: Option<HashMap<String, String>>,
    ) -> Result<FetchResponse, EngineError>;

    /// Get fetcher statistics
    fn statistics(&self) -> FetcherStatistics;
}

/// Standard interface for caches
#[async_trait]
pub trait Cache<K, V>: Send + Sync {
    /// Get value from cache
    async fn get(&self, key: &K) -> Result<Option<V>, EngineError>;

    /// Put value in cache
    async fn put(&self, key: K, value: V, ttl_seconds: Option<u64>) -> Result<(), EngineError>;

    /// Remove value from cache
    async fn remove(&self, key: &K) -> Result<(), EngineError>;

    /// Clear all cache entries
    async fn clear(&self) -> Result<(), EngineError>;

    /// Get cache statistics
    fn statistics(&self) -> CacheStatistics;
}

/// Standard interface for storage backends
#[async_trait]
pub trait Storage: Send + Sync {
    /// Store data
    async fn store(&self, key: &str, data: &[u8]) -> Result<(), EngineError>;

    /// Retrieve data
    async fn retrieve(&self, key: &str) -> Result<Option<Vec<u8>>, EngineError>;

    /// Delete data
    async fn delete(&self, key: &str) -> Result<(), EngineError>;

    /// List keys with prefix
    async fn list_keys(&self, prefix: &str) -> Result<Vec<String>, EngineError>;

    /// Get storage statistics
    fn statistics(&self) -> StorageStatistics;
}

/// Standard interface for configuration providers
#[async_trait]
pub trait ConfigProvider: Send + Sync {
    /// Get configuration value
    async fn get(&self, key: &str) -> Result<Option<serde_json::Value>, EngineError>;

    /// Set configuration value
    async fn set(&self, key: &str, value: serde_json::Value) -> Result<(), EngineError>;

    /// Watch configuration changes
    async fn watch(
        &self,
        key: &str,
    ) -> Result<tokio::sync::broadcast::Receiver<ConfigChangeEvent>, EngineError>;
}

/// Standard interface for health monitoring
#[async_trait]
pub trait HealthMonitor: Send + Sync {
    /// Record a successful operation
    async fn record_success(
        &self,
        operation: &str,
        duration_ms: u64,
        metadata: Option<HashMap<String, String>>,
    ) -> Result<(), EngineError>;

    /// Record a failed operation
    async fn record_failure(
        &self,
        operation: &str,
        error: &EngineError,
        metadata: Option<HashMap<String, String>>,
    ) -> Result<(), EngineError>;

    /// Get health status
    async fn health_status(&self) -> Result<HealthStatus, EngineError>;

    /// Get health statistics
    async fn statistics(&self) -> Result<HealthStatistics, EngineError>;
}

/// Standard interface for metrics collection
#[async_trait]
pub trait MetricsCollector: Send + Sync {
    /// Record a counter metric
    async fn increment_counter(
        &self,
        name: &str,
        value: u64,
        labels: Option<HashMap<String, String>>,
    ) -> Result<(), EngineError>;

    /// Record a gauge metric
    async fn set_gauge(
        &self,
        name: &str,
        value: f64,
        labels: Option<HashMap<String, String>>,
    ) -> Result<(), EngineError>;

    /// Record a histogram metric
    async fn record_histogram(
        &self,
        name: &str,
        value: f64,
        labels: Option<HashMap<String, String>>,
    ) -> Result<(), EngineError>;

    /// Get collected metrics
    async fn collect(&self) -> Result<HashMap<String, MetricValue>, EngineError>;
}

// ===== Data Structures =====

/// Engine health status
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngineHealthStatus {
    pub status: HealthState,
    pub last_check: chrono::DateTime<chrono::Utc>,
    pub response_time_ms: Option<u64>,
    pub error_count: u64,
    pub success_count: u64,
}

/// Engine statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngineStatistics {
    pub total_requests: u64,
    pub successful_requests: u64,
    pub failed_requests: u64,
    pub average_response_time_ms: f64,
    pub uptime_seconds: u64,
    pub memory_usage_bytes: Option<u64>,
}

/// Fetcher statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FetcherStatistics {
    pub total_requests: u64,
    pub successful_requests: u64,
    pub failed_requests: u64,
    pub total_bytes_downloaded: u64,
    pub average_response_time_ms: f64,
    pub active_connections: u32,
}

/// Cache statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheStatistics {
    pub total_entries: u64,
    pub hit_count: u64,
    pub miss_count: u64,
    pub eviction_count: u64,
    pub hit_rate: f64,
    pub total_size_bytes: u64,
}

/// Storage statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageStatistics {
    pub total_files: u64,
    pub total_size_bytes: u64,
    pub read_operations: u64,
    pub write_operations: u64,
    pub delete_operations: u64,
    pub average_read_time_ms: f64,
    pub average_write_time_ms: f64,
}

/// Configuration change event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfigChangeEvent {
    pub key: String,
    pub old_value: Option<serde_json::Value>,
    pub new_value: serde_json::Value,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

/// Health state enumeration
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum HealthState {
    Healthy,
    Degraded,
    Unhealthy,
    Unknown,
}

/// Overall health status
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthStatus {
    pub overall_state: HealthState,
    pub components: HashMap<String, ComponentHealth>,
    pub last_check: chrono::DateTime<chrono::Utc>,
}

/// Component health status
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComponentHealth {
    pub state: HealthState,
    pub response_time_ms: Option<u64>,
    pub error_message: Option<String>,
    pub last_success: Option<chrono::DateTime<chrono::Utc>>,
    pub last_failure: Option<chrono::DateTime<chrono::Utc>>,
}

/// Health statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthStatistics {
    pub total_checks: u64,
    pub successful_checks: u64,
    pub failed_checks: u64,
    pub average_response_time_ms: f64,
    pub uptime_percentage: f64,
}

/// Metric value types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum MetricValue {
    Counter(u64),
    Gauge(f64),
    Histogram(Vec<f64>),
}

// ===== Default Implementations =====

impl Default for EngineHealthStatus {
    fn default() -> Self {
        Self {
            status: HealthState::Unknown,
            last_check: chrono::Utc::now(),
            response_time_ms: None,
            error_count: 0,
            success_count: 0,
        }
    }
}

impl Default for EngineStatistics {
    fn default() -> Self {
        Self {
            total_requests: 0,
            successful_requests: 0,
            failed_requests: 0,
            average_response_time_ms: 0.0,
            uptime_seconds: 0,
            memory_usage_bytes: None,
        }
    }
}
